### DISCUSSION TITLE
Selectively disable auto-update

### DISCUSSION THREAD

**Discussion by Christilut** [hydraulic-software/conveyor#109] [Ideas]
I asked this a while ago on Discord but I don't think there was a good solution yet.

I'd like to be able to turn off auto-updates (on aggressive mode) selectively. Sometimes my users want to go back a version because of a change in the latest version and while I prefer all my users to always be on the latest and greatest, sometimes they just need a temporary downgrade to fix their problem right now.

Now with auto updates on aggressive mode, users are always updated to the latest version when they start it. I can send them an MSIX of an older version but then they don't have auto updates at all (and maybe it has other implications, it's not the normal way to install when using conveyor). For Mac, I don't have any option to send an older version since it will immediately auto update on start.

My suggestion would be to add some kind of magic string to the installer filename that will let conveyor know what to do. That way we can always add it when needed and don't need to set it on build time.

Here are some example ideas with an app called MyApp, using Windows installer:

`MyApp installer.exe`: Normal behavior
`MyApp installer [NO UPDATE].exe`: No updates at all
`MyApp installer [SKIP UPDATE].exe`: Skips the currently found version and updates when a higher version is found
`MyApp installer [VERSION 1.0.5].exe`: MyApp doesn't update until a higher version number than 1.0.5 is found
`MyApp installer [VERSION 1.x].exe`: MyApp doesn't update until 2.0 or higher is found
`MyApp installer [VERSION 1.1.x].exe`: MyApp doesn't update until 1.2 or 2.0 or higher is found

I think especially `SKIP UPDATE` will be useful since you could put it on any release and still make sure users auto updates once a new version is released.

**Comment by mikehearn** [comment]
Yes this is the downside of the web-style update model. You really have to ensure there are either no regressions or they are fixed immediately as you would with a website, which can be quite challenging sometimes.

Giving the user an old MSIX won't work, it'll still update. An old zip release would work.

Putting update constraints into the app name is an interesting hack. I guess there could also be an approach like "hold shift whilst running the installer" to get a GUI that lets you pick the version you want. Or it could be exposed via the upcoming control API and it's up to you to supply UI in the app itself.

**Reply by Christilut** [reply]
A GUI would be super cool but much more work and more fragile. Filename text is definitely a hack but could be a good short term solution.

I'm not sure how it would work with the API, if the user starts the app you should already have decided if it should update right? So sending them an old version with a popup inside the app wouldn't be reached in time.

A GUI would be the ultimate solution I think, user friendly and easily triggered without creating a new installer when the GUI is desired. On Mac it's common to show a popup when holding OPTION when starting an app.

**Reply by mikehearn** [reply]
Conveyor 15 has added an API allowing your app to take more control over the update process. Check it out, maybe it'll suit your needs:

https://conveyor.hydraulic.dev/15.0/release-notes/
